File: BackEnd/model/Vocabulary.js
import mongoose from 'mongoose';

const VocabularySchema = new mongoose.Schema({
    word: { type: String, required: true, index: true }, // Kanji: 学生
    hiragana: { type: String, required: true },          // Kana: がくせい
    meaning: { type: String, required: true },           // Nghĩa: Học sinh
    level: { type: String, enum: ['N5', 'N4', 'N3', 'N2', 'N1'], index: true },
    // tinh huống sử dụng từ
    usage_context: { type: String }, 
    // Media
    audio_url: String,
    image_url: String,
    // Bài học chứa từ này — **không bắt buộc**.
    //
    // Trước đây `required: true`, nghĩa là không thể nhập một kho từ vựng
    // trước khi chia bài. Nhưng từ vựng là thực thể độc lập: một từ có thể
    // xuất hiện ở nhiều bài, hoặc chưa thuộc bài nào, và tiến độ SRS bám vào
    // *từ* chứ không bám vào bài. Bắt buộc bài học ở đây buộc mọi lần nhập
    // phải bịa ra một bài giữ chỗ, và mỗi lần chia lại bài là một lần tạo bản
    // ghi mới — mất sạch tiến độ học của từ đó.
    lesson: { type: mongoose.Schema.Types.ObjectId, ref: 'Lesson', index: true },
    // Ví dụ (Quan trọng)
    examples: [{
        sentence: String, // 私は学生です
        meaning: String,  // Tôi là học sinh
        audio_url: String
    }],

    // Link tới Kanji trong từ này
    related_kanjis: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Kanji' }]
});

// Index text để search nhanh
VocabularySchema.index({ word: 'text', meaning: 'text' });

// Khoá tự nhiên của một từ: mặt chữ **và** cách đọc.
//
// Chỉ `word` là không đủ — tiếng Nhật có từ đồng hình khác âm khác nghĩa
// (今日 = きょう "hôm nay" và こんにち "thời nay"), gộp chúng làm một là sai
// dữ liệu. Ngược lại, thiếu ràng buộc này thì nhập lại cùng một file sẽ sinh
// bản sao, và đó chính là thứ khiến công cụ nhập cũ phải `deleteMany` toàn
// bộ collection trước khi ghi — cách làm đổi `_id` của mọi từ, kéo theo mất
// tiến độ SRS và mất `LessonProgress`.
//
// Có ràng buộc này thì công cụ nhập upsert được theo khoá tự nhiên: chạy lại
// bao nhiêu lần cũng ra đúng một bản ghi, `_id` không đổi.
VocabularySchema.index({ word: 1, hiragana: 1 }, { unique: true });

export default mongoose.model('Vocabulary', VocabularySchema);

File: FrontEnd/lib/models/vocabulary.dart
class Vocabulary {
  final String id;
  final String word;          // Kanji: 学生
  final String hiragana;      // Kana: がくせい
  final String meaning;       // Nghĩa: Học sinh
  final String? level;        // N5, N4, N3, N2, N1
  final String? usageContext; // Tình huống sử dụng
  final String? audioUrl;
  final String? imageUrl;
  final String? lessonId;
  final List<VocabExample> examples;
  final List<String> relatedKanjis;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Vocabulary({
    required this.id,
    required this.word,
    required this.hiragana,
    required this.meaning,
    this.level,
    this.usageContext,
    this.audioUrl,
    this.imageUrl,
    this.lessonId,
    this.examples = const [],
    this.relatedKanjis = const [],
    this.createdAt,
    this.updatedAt,
  });

  factory Vocabulary.fromJson(Map<String, dynamic> json) {
    return Vocabulary(
      id: json['_id'] ?? '',
      word: json['word'] ?? '',
      hiragana: json['hiragana'] ?? '',
      meaning: json['meaning'] ?? '',
      level: json['level'],
      usageContext: json['usage_context'],
      audioUrl: json['audio_url'],
      imageUrl: json['image_url'],
      lessonId: json['lesson'] is Map ? json['lesson']['_id'] : json['lesson'],
      examples: (json['examples'] as List<dynamic>?)
              ?.map((e) => VocabExample.fromJson(e))
              .toList() ??
          [],
      relatedKanjis: (json['related_kanjis'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'word': word,
      'hiragana': hiragana,
      'meaning': meaning,
      'level': level,
      'usage_context': usageContext,
      'audio_url': audioUrl,
      'image_url': imageUrl,
      'lesson': lessonId,
      'examples': examples.map((e) => e.toJson()).toList(),
      'related_kanjis': relatedKanjis,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  Vocabulary copyWith({
    String? id,
    String? word,
    String? hiragana,
    String? meaning,
    String? level,
    String? usageContext,
    String? audioUrl,
    String? imageUrl,
    String? lessonId,
    List<VocabExample>? examples,
    List<String>? relatedKanjis,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Vocabulary(
      id: id ?? this.id,
      word: word ?? this.word,
      hiragana: hiragana ?? this.hiragana,
      meaning: meaning ?? this.meaning,
      level: level ?? this.level,
      usageContext: usageContext ?? this.usageContext,
      audioUrl: audioUrl ?? this.audioUrl,
      imageUrl: imageUrl ?? this.imageUrl,
      lessonId: lessonId ?? this.lessonId,
      examples: examples ?? this.examples,
      relatedKanjis: relatedKanjis ?? this.relatedKanjis,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class VocabExample {
  final String sentence;  // 私は学生です
  final String meaning;   // Tôi là học sinh
  final String? audioUrl;

  VocabExample({
    required this.sentence,
    required this.meaning,
    this.audioUrl,
  });

  factory VocabExample.fromJson(Map<String, dynamic> json) {
    return VocabExample(
      sentence: json['sentence'] ?? '',
      meaning: json['meaning'] ?? '',
      audioUrl: json['audio_url'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'sentence': sentence,
      'meaning': meaning,
      'audio_url': audioUrl,
    };
  }
}
